server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java

```diff
@@ -229,6 +235,93 @@ public void testNotifiesOnFailure() throws InterruptedException {
             }
         }
 
-        assertTrue(countDownLatch.await(10, TimeUnit.SECONDS)); // i.e. it doesn't leak any listeners
+        safeAwait(countDownLatch); // i.e. it doesn't leak any listeners
+    }
+
+    @TestLogging(reason = "testing log output", value = "org.elasticsearch.cluster.routing.BatchedRerouteService:DEBUG")
+    public void testExceptionFidelity() {
+
+        final var mockLogAppender = new MockLogAppender();
+        try (var ignored = mockLogAppender.capturing(BatchedRerouteService.class)) {
+
+            clusterService.getMasterService()
+                .setClusterStatePublisher(
+                    (event, publishListener, ackListener) -> publishListener.onFailure(new FailedToCommitClusterStateException("simulated"))
+                );
+
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "unexpected failure",
+                    BatchedRerouteService.class.getCanonicalName(),
+                    Level.ERROR,
+                    "unexpected failure"
+                )
+            );
+
+            final BatchedRerouteService failingRerouteService = new BatchedRerouteService(clusterService, (s, r, l) -> {
+                throw new ElasticsearchException("simulated");
+            });
+            final var rerouteFailureFuture = new PlainActionFuture<Void>();
+            failingRerouteService.reroute("publish failure", randomFrom(EnumSet.allOf(Priority.class)), rerouteFailureFuture);
+            assertThat(
+                expectThrows(ExecutionException.class, ElasticsearchException.class, () -> rerouteFailureFuture.get(10, TimeUnit.SECONDS))
+                    .getMessage(),
+                equalTo("simulated")
+            );
+            mockLogAppender.assertAllExpectationsMatched();
+
+            final BatchedRerouteService batchedRerouteService = new BatchedRerouteService(clusterService, (s, r, l) -> {
```

Comment:
`// Second case: assert publish failure results in an unwrapped FailedToCommitClusterStateException`